sdk/core/core/inc/az_json_string.h

```diff
@@ -15,6 +15,24 @@
 AZ_NODISCARD az_result
 az_span_reader_read_json_string_char(az_span_reader * const self, uint32_t * const out);
 
+/**
+ * Encodes the given character into a JSON escape sequence. The function returns an empty span if
+ * the given character doesn't require to be escaped.
+ */
+AZ_NODISCARD az_span az_json_esc_encode(az_result_byte const c);
+
+enum {
+  AZ_HEX_LOWER_OFFSET = 'a' - 10,
+  AZ_HEX_UPPER_OFFSET = 'A' - 10,
+};
```

Comment:
???
Where is the lower offset used? what does 'a' - 10 is 'W', 'A' - 10 is '7', what??? :)
Of course I understand how it ends up working, but it is confusing and it is not the most straightforward way to implement.
Enum is public, but it is not useful.
I would suggest to implement in in a way I implemented it in base64.c:
```
AZ_NODISCARD AZ_INLINE uint8_t int4_to_hex(uint8_t const int4) {
  assert(int4 <= 0x0F);
  return (int4 < 10) ? ('0' + int4) : ('A' + (int4 - 10));
}
```

Reply:
@antkmsft, @gilbertw, @vhvb1989 I have conflicting proposals here. There are several problems:
1. If inside inline function I would like to use something that is private, how can I do this? I think, the answer should be provided be in this issue #216 
2. premature optimization, when we use constants assuming that they are better then a constant expression, like `('A' - 10)`.

Reply:
My opinion: based on az_assumptions.c, 'A', is a number, it is just another way to write 65*. I think there should be no constant in this case, the same reason you con't create constants like ONE = 1, TWO = 2.
We would make 1 a constant i one case - if it is a point of configuration, something that we may change.
In the context of converting to hex, it is '0'..'9' 'A' ..'F'/'a'..'f' by definition, it is never going to change on planet earth. Maybe if Martians have their alphabet arranged in different order - then yes, but not until then.
--
* - As to why we only support compiling on ASCII-order for latin alphabet: we assume/only work correctly on ASCII/UTF8, because all the JSON responses we are going to get from azure are going to be ASCII/UTF8; and if we want to compile on EBCDIC, for example - there will be far more places you would need to make changes and write additional code, other than to remap 'A' to something else. Plus, EBCDIC is long dead, and UTF is here to stay, so probably will hever gonna happen